### ISSUE TITLE
i18n fallback in SSG mode does not generate pages

### ISSUE THREAD

**Issue by adamkoch** [withastro/astro#10957]
### Astro Info

```block
My machine:
Astro                    v4.7.0
Node                     v18.18.2
System                   macOS (arm64)
Package Manager          npm
Output                   static
Adapter                  none
Integrations             @astrojs/tailwind
                         @astrojs/mdx
                         @astrojs/sitemap
                         @astrojs/react
                         astro-icon


Stackblitz machine:
Astro                    v4.7.1
Node                     v18.18.0
System                   Linux (x64)
Package Manager          unknown
Output                   static
Adapter                  none
Integrations             none
```

### Describe the Bug

This may just be me misunderstanding how [i18n fallback](https://docs.astro.build/en/guides/internationalization/#fallback) works in SSG mode, but when i18n fallback is configured, no fallback is created automatically.

With the example in the docs:
```
export default defineConfig({
  i18n: {
    defaultLocale: "en",
    locales: ["es", "en", "fr"],
    fallback: {
      fr: "es"
    }
  }
})
```

The docs say:
```
Astro will ensure that a page is built in src/pages/fr for every page that exists in src/pages/es/. If the page does not already exist, then a page with a redirect to the corresponding es route will be created.
```

When testing this and going to `/fr` where there is no page `/fr/index.astro`, it shows shows a 404.

I asked kapa.ai in the discord and [it seemed to agree](https://discord.com/channels/830184174198718474/1236823110691328060) it should be created without needing to set up dynamic routes for locale along with `getStaticPaths()`.

### What's the expected result?

Loading `localhost/fr` should render the `/pages/index.astro` file with `currentLocale` set to `fr` (even when no `/pages/fr/index.astro` exists).

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-qffjm2

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by ematipico** [comment]
I can see the bug, but your expectations isn't correct:

> Loading localhost/fr should render the /pages/index.astro file with currentLocale set to fr (even when no /pages/fr/index.astro exists).

`localhost/fr` will **redirect** to `localhost`. There's no rendering here. The docs are explicit about it:

> you can redirect a user from one locale to another on a per-language basis.

**Comment by adamkoch** [comment]
> localhost/fr will redirect to localhost. There's no rendering here. The docs are explicit about it:

Fair enough but a redirect also doesn't seem to be working. The docs do say:
> Astro will ensure that a page is built in src/pages/fr for every page that exists in src/pages/es/.

But I guess what you're saying is the page is built but all it has in it is a redirect to the fallback? I guess a feature request would be to be able to configure that behavior (perhaps possible with `manual` mode).

What I don't understand then, is if I have translated pages for all my site, but a single page not translated, and a user gets *redirected* to another locale as a fallback. Won't Astro then set `currentLocale` to the fallback locale, and therefore all future links from that page will also be the fallback locale? It seems like a better set up would be to render the fallback locale but with `currentLocale` still set to the original locale, but with some other property set which indicates it's a fallback page (so a banner or message can be displayed).

> Also, your example isn't correct. Can you fix it? You created an en/ folder but didn't set prefixDefaultLocale: true.

Apologies, [updated](https://stackblitz.com/edit/github-qffjm2), I deleted the `en/` folder as for my set up I had `prefixDefaultLocale: false`.

Thanks!

**Comment by PhilStar101** [comment]
Same here! Any updates?

**Comment by PhilStar101** [comment]
I've looked at the code, apparently the astro i18n middleware just handles it using `context.redirect()`. But `redirect()` only works in **SSR mode**. I think the docs don't describe this behavior very clearly and is misleading.

**Comment by PhilStar101** [comment]
It would be nice if middleware will create some stub pages for **SSG mode**. Like [here](https://docs.astro.build/en/guides/routing/#configured-redirects) using [`meta refresh`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/meta#examples)

**Comment by ematipico** [comment]
>  But redirect() only works in SSR mode. I think the docs don't describe this behavior very clearly and is misleading.

That's incorrect. `redirect()` works in SSG mode and it taps the redirect system that Astro has in place: https://docs.astro.build/en/guides/routing/#dynamic-redirects

In SSG, it will create a page with the meta tags you linked.

**Comment by ematipico** [comment]
@adamkoch 

> Won't Astro then set currentLocale to the fallback locale, and therefore all future links from that page will also be the fallback locale? It seems like a better set up would be to render the fallback locale but with currentLocale still set to the original locale, but with some other property set which indicates it's a fallback page (so a banner or message can be displayed).

I believe this is also an incorrect expectation. `currentLocale` is computed from the URL https://docs.astro.build/en/reference/api-reference/#astrocurrentlocale

> The locale computed from the current URL, using the syntax specified in your locales configuration. If the URL does not contain a /[locale]/ prefix, then the value will default to i18n.defaultLocale.

I am aware that users requested a way to programmatically flag a route with a specific locale, however, I haven't seen a concrete proposal. Plus, such proposal would require cookies, in case this information needs to be saved somewhere.

**Comment by adamkoch** [comment]
> That's incorrect. `redirect()` works in SSG mode and it taps the redirect system that Astro has in place: https://docs.astro.build/en/guides/routing/#dynamic-redirects
> 
> In SSG, it will create a page with the meta tags you linked.

But does the current i18n middleware actually use redirect() for fallbacks? As per the stackblitz I created, if you load `/fr` it just shows a 404 when it should redirect to `/`.

> I believe this is also an incorrect expectation. currentLocale is computed from the URL https://docs.astro.build/en/reference/api-reference/#astrocurrentlocale

That's my point exactly (that it's computed from the URL), so if a user hits an i18n fallback (and `redirect()` actually works) then it would result in the locale being reset to the default locale when ideally you would want to render the fallback page but keep the current locale.

**Comment by PhilStar101** [comment]
> That's incorrect. `redirect()` works in SSG mode and it taps the redirect system that Astro has in place: https://docs.astro.build/en/guides/routing/#dynamic-redirects
> 
> In SSG, it will create a page with the meta tags you linked.

i18n middleware uses `context.redirect()` for its redirects but you refer to [`Astro.redirect()`](https://docs.astro.build/en/reference/api-reference/#contextredirect) it is different methods. Also docs about `context.redirect()` [are very clear](https://docs.astro.build/en/reference/api-reference/#contextredirect) that it is only available in **SSR Mode**

**Comment by ematipico** [comment]
>  As per the stackblitz I created, if you load /fr it just shows a 404 when it should redirect to /.

That's the bug we will try to fix :)

**Comment by adamkoch** [comment]
> That's the bug we will try to fix :)

👍 

> I am not really sure what you're talking about here, but I suppose that's not connected to the original issue

Sorry, it's quite hard to explain. It is related though and would be an alternative way to solve this issue. I will try again.

The way I understand i18n fallback, is that it's useful for the case you have a primary language (say English), and then you have translated 90% of your site into other languages (say, FR and ES), but you potentially missed a page or two and instead of showing a 404, you would prefer to have it render the default language (EN).

So, a user who wants FR comes along, and is browsing your site in FR locale and finding all the FR pages. Now they load the page that has no FR version, instead, the fallback EN version will be redirected to (once the issue is fixed) and the user at least gets the page in EN. However, once this page loads, the locale (detected from the URL) will now be EN. Therefore all links on that page will now be linking to the EN paths. So if this FR user clicks to another page which does have a FR version, it won't load it anymore, it will instead load the EN versions going forward.

I'm not sure if this is WAI, it may well be. But the alternative would be to NOT use a redirect for the fallback and instead generate a duplicate version of the EN fallback page but at the FR path.

For SSG, the difference would be like this:

Solving the issue via `redirect()`:
 - `/fr` path would generate an empty HTML page with meta refresh tag point to `/en`

Solving the issue via duplicate EN version:
 - `/fr` path would use the .astro file for `/en` to generate another version for `/fr` but with locale set to FR - this would result in English text/page but with any links on the page still linking to the other FR pages on the site

**Comment by PhilStar101** [comment]
@adamkoch I solved it by doing it by hand. I check posts in my content collection that don't have corresponding translation and add "virtual posts" in `getStaticPaths` for target locale with content of original locale post content + add property fallbackFrom: string

and when rendering post content if there a fallbackFrom prop i display a banner such as  
```ts
`This page has not been translated into ${target locale} at the moment, the content below is in ${fallbackFrom}` 
```

Works great and maybe this solution is even better for UI/UX cause we explicitly tell a user that content isn't available in current language

**Comment by adamkoch** [comment]
Thanks @PhilStar101. That makes sense. I am doing something similar right now. For me, it means adding a `getStaticPaths` to every single page (although I can define the method one time in a central file). It does work fine, but doesn't seem very elegant. For example, if another team member was to add a new page they may forget to add the defined `getStaticPaths`. I guess another solution would be to have some sort of global `getStaticPaths` setter to not have to do it for every page.

The fallback banner is nice though. That could theoretically be provided by i18n as well though by setting some value if the fallback.

**Comment by ematipico** [comment]
@adamkoch 

The feature that you asked for wasn't available, that's why I reduced the scope to redirects. Soon, we will stabilise the `Astro.rewrite` feature, which will do what you asked (and what I wanted for a long time). We will add a new feature to i18n to choose between redirects and rewrites.

> However, once this page loads, the locale (detected from the URL) will now be EN. Therefore all links on that page will now be linking to the EN paths. So if this FR user clicks to another page which does have a FR version, it won't load it anymore, it will instead load the EN versions going forward.

Feel free to drop a proposal to assign programmatically a local to a route. I did open this thread some months ago: https://github.com/withastro/roadmap/discussions/795

Regarding the issue, I opened the stackblitz, **used the latest of Astro**, and I could see the `/fr/index.html` page correctly generated as a redirect.
